orttraining/orttraining/python/training/ortmodule/_execution_agent.py

```diff
@@ -109,27 +159,103 @@ def __init__(self, path_or_bytes, fw_feed_names, fw_outputs_device_info,
+                        _utils.get_device_index(self._device)
+            )] * len(self._graph_info.user_output_names)
+
+        bw_fetches_names = [output.name for output in onnx_model.graph.output]
+        bw_outputs_device_info = [
+            C.OrtDevice(get_ort_device_type(self._device.type),
+                        C.OrtDevice.default_memory(),
+                        _utils.get_device_index(self._device)
+            )] * len(bw_fetches_names)
         self._training_agent = C_TrainingAgent(self._inference_session._sess, fw_feed_names, fw_outputs_device_info,
                                                bw_fetches_names, bw_outputs_device_info)
 
-    def run_forward(self, feeds, fetches, state):
-        """
-         Compute the forward subgraph for given feeds and fetches.
-         :param feeds: Inputs to the graph run.
-         :param fetches: Outputs of the graph run.
-         :param state: State of the graph that is used for executing partial graph runs.
-        """
+
+
+    def _prepare_forward_inputs(self, inputs):
+        """ Prepare feeds from torch tensors """
+
+        # Assert that the input and model device match
+        _utils._check_same_device(self._device, "Input argument to forward", *inputs)
+        # TODO: Try to reuse the output buffers as some of the output tensors are same sizes,
+        #   especially the backward graph outputs.
+        # REVIEW(codemzs): Consolidate Training Agent with InferenceAgent on C++ side to not
+        # have the need for passing IOBinding.
+        forward_inputs = C.OrtValueVector()
+        forward_inputs.reserve(len(inputs))
+        for input in inputs:
+            forward_inputs.push_back(to_dlpack(input), input.dtype == torch.bool)
+
+        return forward_inputs
+
+
+    def _process_forward_outputs(self, forward_outputs, state):
+        user_outputs = tuple(_utils._ortvalue_to_torch_tensor(forward_output) for forward_output in forward_outputs)
+        output_info = [(output.shape, output.device, output.dtype) for output in user_outputs]
+        run_info = RunStateInfo(state, output_info)
+        # Return user outputs and forward run information
+        return user_outputs, run_info
+
+
+    def forward(self, *inputs):
+        feeds = self._prepare_forward_inputs(inputs)
+        fetches = C.OrtValueVector()
+        state = C.PartialGraphExecutionState()
         self._training_agent.run_forward(feeds, fetches, state)
+        return self._process_forward_outputs(fetches, state)
 
-    def run_backward(self, feeds, fetches, state):
-        """
-         Compute the backward subgraph for given feeds and fetches.
-         :param feeds: Inputs to the graph run.
-         :param fetches: Outputs of the graph run.
-         :param state: State of the graph that is used for executing partial graph runs.
-        """
-        self._training_agent.run_backward(feeds, fetches, state)
+
+    def _prepare_backward_inputs(self, ctx, grad_outputs):
```

Comment:
```suggestion
    def _process_backward_inputs(self, ctx, grad_outputs):
```
Better to make terminology uniform

Reply:
Thank you. Done